fail2ban/tests/fail2banclienttestcase.py

```diff
@@ -0,0 +1,593 @@
+			# start (in new process, using the same python version):
+			cmd = (sys.executable, os.path.join(os.path.join(BIN), SERVER))
+			logSys.debug('Start %s ...', cmd)
+			cmd = cmd + startparams + ("-b",)
+			ret = Utils.executeCmd(cmd, timeout=MAX_WAITTIME, shell=False, output=True)
+			self.assertTrue(len(ret) and ret[0])
+			# wait for server (socket and ready):
+			self._wait_for_srv(tmp, True, startparams=cmd)
+			self.assertLogged("Server ready")
+			self.pruneLog()
+			try:
+				self.assertRaises(ExitException, _exec_server, 
+					(SERVER,) + startparams + ("echo", "TEST-ECHO",))
+				self.assertRaises(FailExitException, _exec_server, 
+					(SERVER,) + startparams + ("~~unknown~cmd~failed~~",))
+			finally:
+				self.pruneLog()
+				# stop:
+				self.assertRaises(ExitException, _exec_server, 
+					(SERVER,) + startparams + ("stop",))
+				self.assertLogged("Shutdown successful")
+				self.assertLogged("Exit with code 0")
+		finally:
+			_kill_srv(tmp)
+
+	def _testServerStartForeground(self, tmp, startparams, phase):
+		# start and wait to end (foreground):
+		logSys.debug("-- start of test worker")
+		phase['start'] = True
+		self.assertRaises(fail2bancmdline.ExitException, _exec_server, 
+			(SERVER, "-f") + startparams + ("start",))
+		# end :
+		phase['end'] = True
+		logSys.debug("-- end of test worker")
+
+	@withtmpdir
+	def testServerStartForeground(self, tmp):
+		th = None
+		try:
+			# started directly here, so prevent overwrite test cases logger with "INHERITED"
+			startparams = _start_params(tmp, logtarget="INHERITED")
+			# because foreground block execution - start it in thread:
+			phase = dict()
+			th = Thread(name="_TestCaseWorker", 
+				target=Fail2banServerTest._testServerStartForeground, args=(self, tmp, startparams, phase))
+			th.daemon = True
+			th.start()
+			try:
+				# wait for start thread:
+				Utils.wait_for(lambda: phase.get('start', None) is not None, MAX_WAITTIME)
+				self.assertTrue(phase.get('start', None))
+				# wait for server (socket and ready):
+				self._wait_for_srv(tmp, True, startparams=startparams)
+				self.pruneLog()
+				# several commands to server:
+				self.assertRaises(ExitException, _exec_server, 
+					(SERVER,) + startparams + ("ping",))
+				self.assertRaises(FailExitException, _exec_server, 
+					(SERVER,) + startparams + ("~~unknown~cmd~failed~~",))
+				self.assertRaises(ExitException, _exec_server, 
+					(SERVER,) + startparams + ("echo", "TEST-ECHO",))
+			finally:
+				self.pruneLog()
+				# stop:
+				self.assertRaises(ExitException, _exec_server, 
+					(SERVER,) + startparams + ("stop",))
+				# wait for end:
+				Utils.wait_for(lambda: phase.get('end', None) is not None, MAX_WAITTIME)
+				self.assertTrue(phase.get('end', None))
+				self.assertLogged("Shutdown successful", "Exiting Fail2ban")
+		finally:
+			_kill_srv(tmp)
+			if th:
+				th.join()
+
+	@withtmpdir
+	def testServerFailStart(self, tmp):
+		try:
```

Comment:
ok -- https://github.com/sebres/fail2ban/pull/5   where I called it `@with_kill_srv`.